File: src/archiver.ts
import { createWriteStream, statSync } from "node:fs";
import { readdir, rename, stat, unlink, writeFile } from "node:fs/promises";
import path from "node:path";
import SftpClient from "ssh2-sftp-client";
import constants from "./constants.ts";
import { formatBytes, formatKnownError, logSuccess, logWarning } from "./log/logger.ts";
import msg, { type MessageParams } from "./log/messages.ts";
import { ArchiveDescription } from "./types.ts";
import { findOldArchives, isArchiveFile } from "./utils.ts";
import { writeArchive } from "./writer.ts";

export async function createLocalArchive(
    sources: string[],
    destination: string,
    archive: ArchiveDescription,
    keep: number,
) {
    if (!statSync(destination, { throwIfNoEntry: false })?.isDirectory()) {
        throw new Error(
            msg.get("err.noneExistDestination", {
                path: destination,
            }),
        );
    }

    const archivePath = path.join(destination, archive.fullFilename);
    const checksumPath = path.join(destination, archive.fullChecksumFilename);
    const partialPath = path.join(destination, archive.fullPartialFilename);

    try {
        const archiveData = await writeArchive(
            sources,
            createWriteStream(partialPath),
            archive.compression,
            (entry) => {
                if (path.dirname(entry.path) !== path.resolve(destination)) return false;

                // Exclude the created archive files to prevent them from being included in backups,
                // which could otherwise cause duplicates or recursive backup loops.
                const name = path.basename(entry.path);
                return (
                    isArchiveFile(name, archive.prefix, archive.extension) ||
                    isArchiveFile(name, archive.prefix, archive.checksumExtension) ||
                    isArchiveFile(name, archive.prefix, archive.partialExtension)
                );
            },
        );

        if ((await stat(partialPath))?.size !== archiveData.size) {
            throw new Error(
                msg.get("err.sizeMismatch", {
                    path: partialPath,
                }),
            );
        }

        await rename(partialPath, archivePath);
        await writeFile(checksumPath, `${archiveData.hash}  ${archive.fullFilename}\n`);

        logSuccess(
            msg.get("success.archiveCreated", {
                path: archivePath,
                size: formatBytes(archiveData.size),
            }),
        );

        if (archiveData.warnings > 0) process.exitCode = constants.exitCodes.incomplete;
    } catch (error) {
        await unlink(partialPath).catch(() => {
            logWarning(
                msg.get("warn.deleteFailed", {
                    path: partialPath,
                }),
            );
        });
        await unlink(archivePath).catch(() => {
            logWarning(
                msg.get("warn.deleteFailed", {
                    path: archivePath,
                }),
            );
        });
        await unlink(checksumPath).catch(() => {
            logWarning(
                msg.get("warn.deleteFailed", {
                    path: checksumPath,
                }),
            );
        });
        throw error;
    }

    if (keep <= 0) return;

    const fileNames = (await readdir(destination, { withFileTypes: true }))
        ?.filter((x) => x.isFile())
        ?.map((x) => x.name);

    for (const name of findOldArchives(fileNames, archive.prefix, archive.extension, keep)) {
        const filePath = path.join(destination, name);
        const checksumFilePath = filePath + archive.checksumExtension;

        let failed = false;

        if (fileNames.includes(name))
            await unlink(filePath).catch((error) => {
                failed = true;

                const params: MessageParams = {
                    path: filePath,
                };
                logWarning(formatKnownError(error, params) ?? msg.get("warn.deleteFailed", params));
            });

        if (fileNames.includes(name + archive.checksumExtension))
            await unlink(checksumFilePath).catch((error) => {
                failed = true;

                const params: MessageParams = {
                    path: checksumFilePath,
                };
                logWarning(formatKnownError(error, params) ?? msg.get("warn.deleteFailed", params));
            });

        if (failed) continue;

        logSuccess(
            msg.get("success.archiveDeleted", {
                path: filePath,
            }),
        );
    }
}

export async function createSftpArchive(
    sources: string[],
    destination: string,
    archive: ArchiveDescription,
    keep: number,
    sftpConfig: SftpClient.ConnectOptions,
) {
    const archivePath = path.posix.join(destination, archive.fullFilename);
    const checksumPath = path.posix.join(destination, archive.fullChecksumFilename);
    const partialPath = path.posix.join(destination, archive.fullPartialFilename);

    const sftp = new SftpClient();

    try {
        await sftp.connect(sftpConfig);

        if ((await sftp.exists(destination)) !== "d") {
            throw new Error(
                msg.get("err.noneExistDestination", {
                    path: destination,
                }),
            );
        }

        try {
            const archiveData = await writeArchive(sources, sftp.createWriteStream(partialPath), archive.compression);

            if ((await sftp.stat(partialPath))?.size !== archiveData.size) {
                throw new Error(
                    msg.get("err.sizeMismatch", {
                        path: partialPath,
                    }),
                );
            }

            await sftp.rename(partialPath, archivePath);
            await sftp.put(Buffer.from(`${archiveData.hash}  ${archive.fullFilename}\n`), checksumPath);

            logSuccess(
                msg.get("success.archiveCreated", {
                    path: archivePath,
                    size: formatBytes(archiveData.size),
                }),
            );

            if (archiveData.warnings > 0) process.exitCode = constants.exitCodes.incomplete;
        } catch (error) {
            await sftp.delete(partialPath).catch(() => {
                logWarning(
                    msg.get("warn.deleteFailed", {
                        path: partialPath,
                    }),
                );
            });
            await sftp.delete(archivePath).catch(() => {
                logWarning(
                    msg.get("warn.deleteFailed", {
                        path: archivePath,
                    }),
                );
            });
            await sftp.delete(checksumPath).catch(() => {
                logWarning(
                    msg.get("warn.deleteFailed", {
                        path: checksumPath,
                    }),
                );
            });
            throw error;
        }

        if (keep <= 0) return;

        const fileNames = (await sftp.list(destination))?.filter((x) => x.type === "-")?.map((x) => x.name);

        for (const name of findOldArchives(fileNames, archive.prefix, archive.extension, keep)) {
            const filePath = path.posix.join(destination, name);
            const checksumFilePath = filePath + archive.checksumExtension;

            let failed = false;

            if (fileNames.includes(name))
                await sftp.delete(filePath).catch((error) => {
                    failed = true;

                    const params: MessageParams = {
                        path: filePath,
                    };
                    logWarning(formatKnownError(error, params) ?? msg.get("warn.deleteFailed", params));
                });

            if (fileNames.includes(name + archive.checksumExtension))
                await sftp.delete(checksumFilePath).catch((error) => {
                    failed = true;

                    const params: MessageParams = {
                        path: checksumFilePath,
                    };
                    logWarning(formatKnownError(error, params) ?? msg.get("warn.deleteFailed", params));
                });

            if (failed) continue;

            logSuccess(msg.get("success.archiveDeleted", { path: filePath }));
        }
    } finally {
        await sftp.end().catch(() => {});
    }
}

File: src/anqi/archiver.py
import posixpath
import sys
import tarfile
from pathlib import Path

from paramiko import SFTPClient

from .types import ArchiveDescription, HashingWriter
from .utils import (
    SftpConfig,
    add_to_archive,
    find_old_archives,
)


def create_local_archive(
    sources: list[str],
    destination: Path,
    archive_description: ArchiveDescription,
    keep: int,
):
    archive_path = destination / archive_description.full_filename

    if not archive_path.parent.is_dir():
        print(
            f"Destination {archive_path.parent} is not a directory or does not exist!",
            file=sys.stderr,
        )
        sys.exit(1)

    with archive_path.open("wb") as file, HashingWriter(file) as writer:
        with tarfile.open(fileobj=writer, mode="w:gz") as archive_file:
            add_to_archive(archive_file, sources)

        hash_path = Path(f"{archive_path}.sha256")

        with hash_path.open("w") as hash_file:
            hash_file.write(
                f"{writer.hexdigest()}  {archive_description.full_filename}\n"
            )

        print(f"Backup created at {archive_path}")

    if keep <= 0:
        return

    old_archives = find_old_archives(
        [x.name for x in destination.iterdir() if x.is_file()],
        archive_description.prefix,
        archive_description.extension,
        keep,
    )

    for file_name in old_archives:
        file_path = destination / file_name
        file_path.unlink()

        hash_path = file_path.with_name(f"{file_name}.sha256")
        if hash_path.exists():
            hash_path.unlink()

        print(f"Backup deleted at {file_path}")


def create_sftp_archive(
    sources: list[str],
    destination: str,
    archive_description: ArchiveDescription,
    keep: int,
    sftp_config: SftpConfig,
):
    remote_archive_path = posixpath.join(destination, archive_description.full_filename)

    with (
        sftp_config.connection() as conn,
        conn.sftp() as sftp,
    ):
        sftp: SFTPClient

        with (
            sftp.file(remote_archive_path, "wb") as file,
            HashingWriter(file) as writer,
        ):
            with tarfile.open(fileobj=writer, mode="w|gz") as archive_file:
                add_to_archive(archive_file, sources)

            remote_size = sftp.stat(remote_archive_path).st_size

            if remote_size != writer.size:
                print(
                    f"File size mismatch after upload for {remote_archive_path}",
                    file=sys.stderr,
                )
                sys.exit(1)

            with sftp.file(remote_archive_path + ".sha256", "w") as hash_file:
                hash_file.write(
                    f"{writer.hexdigest()}  {archive_description.full_filename}\n"
                )

            print(f"Backup created at {remote_archive_path}")

        if keep <= 0:
            return

        # files = set(sftp.listdir(destination))
        files = sftp.listdir(destination)
        old_archives = find_old_archives(
            files,
            archive_description.prefix,
            archive_description.extension,
            keep,
        )

        for file_name in old_archives:
            file_path = posixpath.join(destination, file_name)
            sftp.remove(file_path)

            hash_path = f"{file_path}.sha256"
            if f"{file_name}.sha256" in files:
                sftp.remove(hash_path)

            print(f"Backup deleted at {file_path}")
